core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java

```diff
@@ -0,0 +1,605 @@
+        super(metadata);
+        this.base = base;
+        this.from = from;
+        this.to = to;
+        this.stride = stride;
+        this.timestampIndex = timestampIndex;
+        this.values = fillValues;
+        this.metadata = metadata;
+    }
+
+    @Override
+    public RecordCursorFactory getBaseFactory() {
+        return base;
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        if (metadata.getColumnCount() > this.values.size() + 1) {
+
+            if (this.values.size() == 1 && this.values.getQuick(0).isNullConstant()) {
+                final int diff = (metadata.getColumnCount() - 1);
+                // skip one entry as it should be the designated timestamp
+                for (int i = 1; i < diff; i++) {
+                    values.add(NullConstant.NULL);
+                }
+            } else {
+                throw SqlException.$(-1, "not enough fill values");
+            }
+
+        }
+
+        final RecordCursor baseCursor = base.getCursor(executionContext);
+        try {
+            cursor.of(baseCursor, executionContext.getCircuitBreaker(), from, to, stride, values, timestampIndex);
+            return cursor;
+        } catch (Throwable th) {
+            baseCursor.close();
+            throw th;
+        }
+    }
+
+    @Override
+    public RecordMetadata getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+    @Override
+    public void toPlan(PlanSink sink) {
+        sink.type("Fill Range");
+        sink.attr("range").val('(').val(from).val(',').val(to).val(')');
+        sink.attr("stride").val('\'').val(stride).val('\'');
+        sink.attr("values").val(values);
+        sink.child(base);
+    }
+
+    @Override
+    public boolean usesCompiledFilter() {
+        return base.usesCompiledFilter();
+    }
+
+    @Override
+    public boolean usesIndex() {
+        return base.usesIndex();
+    }
+
+    @Override
+    protected void _close() {
+        base.close();
+    }
+
+    private static class FillRangeRecordCursor implements NoRandomAccessRecordCursor {
+        private static final int RANGE_FULLY_BOUND = 1;
+        private static final int RANGE_LOWER_BOUND = 2;
+        private static final int RANGE_UNBOUNDED = 0;
+        private static final int RANGE_UPPER_BOUND = 3;
+        protected final FillRangeRecord fillingRecord = new FillRangeRecord();
+        protected final TimestampFloorOffsetFunctionFactory flooringFactory = new TimestampFloorOffsetFunctionFactory();
+        protected RecordCursor baseCursor;
+        protected Record baseRecord;
+        protected SqlExecutionCircuitBreaker circuitBreaker;
+        protected int fillOffset;
```

Comment:
nit: the name of this field is a bit misleading. It stores a bucket index, but it's called `fillOffset`. Something like `fillBucketIndex` or `currentBucketIndex` or `bucketIndex`. Same with `nextBucket`: it could be `nextBucketTimestamp`